cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java

```diff
@@ -54,29 +55,67 @@ private Schema getSchema(String schemaStr) {
     if (schema == null) {
       return Collections.emptySet();
     }
+    Set<String> indexes = new HashSet<>();
+    processSchema(indexes, schema, null);
+    return indexes;
+  }
+
+  private void processSchema(Set<String> indexes, Schema schema, @Nullable String fieldName) {
+    switch (schema.getType()) {
+      case NULL:
+        // Ignore null types
+        break;
+      case BOOLEAN:
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BYTES:
+      case ENUM:
+      case STRING:
+        createIndexes(indexes, schema, fieldName);
+        break;
+      case ARRAY:
+        createIndexes(indexes, schema, fieldName);
+        processSchema(indexes, schema.getComponentSchema(), schema.getComponentSchema().getRecordName());
+        break;
+      case MAP:
+        createIndexes(indexes, schema, fieldName);
+        Map.Entry<Schema, Schema> mapSchema = schema.getMapSchema();
+        processSchema(indexes, mapSchema.getKey(), mapSchema.getKey().getRecordName());
+        processSchema(indexes, mapSchema.getValue(), mapSchema.getValue().getRecordName());
+        break;
+      case RECORD:
+        createIndexes(indexes, schema, fieldName);
+        for (Schema.Field field : schema.getFields()) {
+          processSchema(indexes, field.getSchema(), field.getName());
+        }
+        break;
+      case UNION:
+        // if something is nullable, it is a union of null and the other type.
+        if (schema.isNullable()) {
+          createIndexes(indexes, schema.getNonNullable(), fieldName);
+        } else {
+          for (Schema us : schema.getUnionSchemas()) {
+            createIndexes(indexes, us, fieldName);
```

Comment:
It produces :UNION, except for a nullable type (a union of type and null) where it produces :type.